### PR TITLE
Add solution for Module-4

### PR DESCRIPTION
**Pull request by cezary-kania**

### CODE REVIEW

File: Module-4/Task2/NumberParser.cs [hunk 1]
```diff
@@ -1,12 +1,38 @@
...
+        }
+        var firstChar = trimmedStringValue[0];
+        var startIndex = char.IsNumber(firstChar) ? 0 : 1;
+        var sign = firstChar is '-' ? -1 : 1;
+        if (startIndex >= stringValue.Length)
+        {
+            throw new FormatException(invalidFormatMessage);
+        }
+        long result = 0;
+        for (var i = startIndex; i < trimmedStringValue.Length; ++i)
```
Comment (KirylPrykhodzka) [inline Module-4/Task2/NumberParser.cs]: why are you using ++i here and not i++ ?

File: Module-4/Task2/NumberParser.cs [hunk 2]
```diff
@@ -1,12 +1,38 @@
...
         {
-            throw new NotImplementedException();
+            if (trimmedStringValue[i] < '0' || trimmedStringValue[i] > '9')
+            {
+                throw new FormatException(invalidFormatMessage);
+            }
+            result = result * 10 + (trimmedStringValue[i] - '0');
+            if (sign * result is > int.MaxValue or < int.MinValue)
```
Comment (KirylPrykhodzka) [inline Module-4/Task2/NumberParser.cs:31]: try using Math.Abs here

File: Module-4/Task3/UserTaskController.cs [hunk 3]
```diff
@@ -1,42 +1,29 @@
...
-                return false;
-            }
-
-            return true;
+            var task = new UserTask(description);
+            _taskService.AddTaskForUser(userId, task);
         }
-
-        private string GetMessageForModel(int userId, string description)
+        catch (Exception ex)
```
Comment (KirylPrykhodzka) [inline Module-4/Task3/UserTaskController.cs]: Let's improve this catch block.
Try to imagine what will happen if an unexpected error occurs in UserTaskService. Do you want to return ex.Message to the user in this case?

**Review by KirylPrykhodzka — CHANGES_REQUESTED** [review]

File: Module-4/Task3/UserTaskController.cs (hunk 3 above)
Comment (cezary-kania) [reply Module-4/Task3/UserTaskController.cs]: I just added new custom exception type. Please check.

File: Module-4/Task2/NumberParser.cs (hunk 1 above)
Comment (cezary-kania) [reply Module-4/Task2/NumberParser.cs]: Done

File: Module-4/Task3/UserTaskController.cs (hunk 3 above)
Comment (KirylPrykhodzka) [reply Module-4/Task3/UserTaskController.cs]: Or you could have added multiple catch blocks for multiple types of exceptions 😄

Comment (KirylPrykhodzka) [reply Module-4/Task3/UserTaskController.cs]: Just for training purposes please add a catch block per exception type. This is the usual way of achieving what we want here (although your solution works, too)

**Review by KirylPrykhodzka — APPROVED** [review]

**Pull request merged**